src/pygenstability/data_clustering.py

```diff
@@ -0,0 +1,286 @@
+"""Construct geometric graphs from data for multiscale clustering."""
+
+import numpy as np
+
+from scipy.spatial.distance import pdist, squareform
+from scipy.sparse.csgraph import minimum_spanning_tree
+from scipy.sparse import csr_matrix
+
+from pygenstability.pygenstability import run as pgs_run
+from pygenstability.plotting import plot_scan as pgs_plot_scan
+from pygenstability.optimal_scales import identify_optimal_scales
+
+
+def compute_kNN(D, k=5):
+    """Computes kNN graph."""
+
+    N = D.shape[0]
+
+    # get k nearest neighbours for each point
+    k_neighbours = np.argsort(D, axis=1)[:, 1 : k + 1]
+
+    # initialise adjacency matrix
+    A = np.zeros((N, N))
+
+    # build kNN graph
+    for i in range(N):
+        for neighbour in k_neighbours[i]:
+            A[i, neighbour] = D[i, neighbour]
+            A[neighbour, i] = A[i, neighbour]
+
+    return A
+
+
+def compute_CkNN(D, k=5, delta=1):
+    """Computes CkNN graph."""
+    # obtain rescaled distance matrix, see CkNN paper
+    darray_n_nbrs = np.partition(D, k)[:, [k]]
+    ratio_matrix = D / np.sqrt(darray_n_nbrs.dot(darray_n_nbrs.T))
+    # threshold rescaled distance matrix by delta
+    A = D * (ratio_matrix < delta)
+    return A
+
+
+class GraphConstruction:
+    """Graph construction."""
+
+    def __init__(
+        self,
+        metric="euclidean",
+        method="cknn",
+        k=5,
+        delta=1.0,
+        distance_threshold=np.inf,
+    ):
+        # parameters
+        self.metric = metric
+        self.method = method
+        self.k = k
+        self.delta = delta
+        self.distance_threshold = distance_threshold
+
+        # attributes
+        self.adjacency_ = csr_matrix
```

Comment:
I would put None instead of the class csr_matrix

Reply:
also, why an _ at the end of adjacency? you could drop that, no? Afaik, it is used to avoid conflict with builtin names, like `pass_` as `pass` is a python command for example

Reply:
I can put None instead. For your other question see my comment below.

Reply:
done